Mentions

adrianq opened the pull request.

adrianq committed:
adding mentions infrastructure

adrianq committed:
improve user interaction and panel

adrianq committed:
add most mentioned users

adrianq committed:
cosmetics

adrianq committed:
cosmetics

adrianq committed:
remove most mentioned users from other users, bug fixing, etc

adrianq committed:
Merge branch 'master_229' into mentions

adrianq committed:
better naming and remove unused code

adrianq committed:
extract multilanguage terms

adrianq commented:
Closes EyeSeeTea/dhis2-core#18

tokland commented on src/manager/AppManager.js:
```diff
@@ -232,11 +235,110 @@ AppManager.prototype.init = function(callbackFn) {
+            ],
+            success: function (response) {
+                var mentionsUser = {}
+                previousResponse.interpretations.map( interpretation => 
```
map is used to collect results: `xsDoubled = [1, 2, 3].map(x => 2*x)`. To perform side-effects within a loop, it's more declarative forEach: `[1, 2, 3].forEach(n => console.log(n))`

adrianq replied:
Done!

tokland commented on src/manager/AppManager.js:
```diff
@@ -232,11 +235,110 @@ AppManager.prototype.init = function(callbackFn) {
+                var mentionsUser = {}
+                previousResponse.interpretations.map( interpretation => 
+                    interpretation.mentions.map(mention => {
+                         if (mention.username in mentionsUser){
```
This saves some lines: `mentionsUser[mention.username] = (mentionsUser[mention.username] || 0) + 1;`

adrianq replied:
Much better yep. Changed it!

tokland commented on src/manager/AppManager.js:
```diff
@@ -232,11 +235,110 @@ AppManager.prototype.init = function(callbackFn) {
+                // console.log(mentionsUser);
+                
+                // Sort by mentions
+                mentionsUser = Object.keys(mentionsUser).sort( function(a,b) {
```
It's a bit confusing to reuse the same variable to hold different values. First `mentionUser` is an object, then it's an array of sorted usernames.

adrianq replied:
I am now using two different names... TBH I am very bad with naming

tokland commented on src/manager/AppManager.js:
```diff
@@ -232,11 +235,110 @@ AppManager.prototype.init = function(callbackFn) {
+
+               
+                // Get real user object (No foreign key in json b)
+                t.mostMentionedUsers = mentionsUser.map(username => {
```
Here it's probably not a problem, we don't have so many users or mentions, but loop within a loop has quadratic behaviour. Tipically you would build an intermediate structure so the operations within the loop are O(1) and the total is a lineal O(n). Something like this (untested):

```
var usersByUsername = {};
t.users.forEach(user => usersByUsername[user.userCredentials.username] = user);
t.mostMentionedUsers = mentionsUser.map(username => usersByUsername[username]);
```
You can create `usersByUsername` is a single line using a `reduce`, but  it's very cryptic, I would not do it. Underscore/lodash have methods (i.e fromPairs()) to make this task very simple.

adrianq replied:
Your code was not tested but it works like a charm :) I also improved a bit the comments

tokland commented on src/ui/EastRegion.js:
```diff
@@ -285,6 +285,96 @@ EastRegion = function(c) {
+            var potentialUsers = appManager.users
+                .filter(user => user.userCredentials.username.includes(currentMention))
+                .map((user) => {
+                    return {
```
Code repeated two times, maybe abstracted in a function?

adrianq replied:
Extracted as a function. Note this is now part of the module MentionToolbar as it is used by EastRegion and InterpretationWindow

tokland commented on src/ui/EastRegion.js:
```diff
@@ -285,6 +285,96 @@ EastRegion = function(c) {
+                    }
+                });
+
+            if (potentialMostMentionedUsers != null && potentialMostMentionedUsers.length > 0){
```
`if (potentialMostMentionedUsers != null` In the general case it's better to check simply that is "truthy", like you do in Python: `if (potentialMostMentionedUsers`. This way it covers also `undefined`.

adrianq replied:
Done!

tokland reviewed (changes requested):
Some general comments apart from code specifics:

 (using AT to avoid mentioning github users)

- Autocomplete is not implemented in the "Write Interpretation" popup.
- Visual: 1) Start writing a comment, 2) Write "[AT]andr", selector with users appear, 3) Click "Clear interpretation". Problem: The selector dialog is not hidden.
- Visual: Write "Hello [AT]android". Now go back with the cursor, when you get to "Hello", the dialog re-appears. It cannot be closed other than selecting the user already mentionded. It does not allow to write a user mention at this point.

adrianq committed:
missing dot

adrianq committed:
add extra paremeter to define offset

adrianq committed:
extract mention toolbar and integrate in interpretation window

adrianq committed:
get potential mention using text between 0 and current cursor position

adrianq committed:
improve performance, better naming, remove logs

adrianq committed:
extract redundant code

adrianq commented:
Re the general comments, I have extracted mentionPanel as a module, MentionToolbar, so it can be instantiated both from the EastPanel and the InterpretationWindow. It took some time but it looks much better now.
I am also using the cursor position in the text area. Before this we always consider the cursor to be at the end of the text and this creates some problems as you noticed.
Finally, I have added some event listeners. These listeners handle any click outside the MentionToolbar. If there is a click anywhere we close the mentiontoolbar (I have tried to mimic the Slack approach as much as possible).

adrianq committed:
fix label creation method
